go/vt/topo/topo_lock.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topo
+
+import (
+	"context"
+	"fmt"
+
+	"vitess.io/vitess/go/trace"
+	"vitess.io/vitess/go/vt/log"
+	"vitess.io/vitess/go/vt/proto/vtrpc"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+// ITopoLock is the interface for a lock that can be used to lock a key in the topology server.
+// The lock is associated with a context and can be unlocked by calling the returned function.
+// Note that we don't need an Unlock method on the interface, as the Lock() function
+// returns a function that can be used to unlock the lock.
+type ITopoLock interface {
+	Lock(ctx context.Context) (context.Context, func(*error), error)
+}
+
+type TopoLock struct {
+	Path string // topo path to lock
+	Name string // name, for logging purposes
+
+	ts *Server
+}
+
+var _ ITopoLock = (*TopoLock)(nil)
+
+func (ts *Server) NewTopoLock(path, name string) *TopoLock {
+	return &TopoLock{
+		ts:   ts,
+		Path: path,
+		Name: name,
+	}
+}
+
+func (tl *TopoLock) String() string {
+	return fmt.Sprintf("TopoLock{Path: %v, Name: %v}", tl.Path, tl.Name)
+}
+
```

Comment:
Much of this code is from our existing `locks.go`. In that file we have repeated code for each type of lock (keyspace, shard) without any generalization. Here we introduce a generationalization for the basic functionality required for the keyspace routing rules lock.

The `TopoLock` can be extended to implement the features used in other locks and get rid of the duplication in a future refactor.